File: ecs-hosted/go/dns/applicationDns.go
package main

import (
	"fmt"
	"strings"

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/route53"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func NewApplicationDns(ctx *pulumi.Context, name string, args *ApplicationDnsArgs, opts ...pulumi.ResourceOption) (*ApplicationDns, error) {
	var resource ApplicationDns

	err := ctx.RegisterComponentResource("pulumi:applicationDns", name, &resource, opts...)
	if err != nil {
		return nil, err
	}

	// create our parented options
	options := append(opts, pulumi.Parent(&resource))
	zone := route53.LookupZoneOutput(ctx, route53.LookupZoneOutputArgs{
		Name: args.ZoneName,
	})

	if err != nil {
		return nil, err
	}

	apiName := args.Domain.ApplyT(func(s string) string {
		return strings.Join([]string{"api", s}, ".")
	}).(pulumi.StringOutput)

	resource.ApiRecord, err = route53.NewRecord(ctx, fmt.Sprintf("%s-api-record", name), &route53.RecordArgs{
		ZoneId: zone.Id(),
		Name:   apiName,
		Type:   pulumi.String("A"),
		Aliases: &route53.RecordAliasArray{
			route53.RecordAliasArgs{
				Name:                 args.PublicLoadBalancerDnsName,
				ZoneId:               args.PublicLoadBalancerZoneId,
				EvaluateTargetHealth: pulumi.Bool(true),
			},
		},
	}, options...)

	if err != nil {
		return nil, err
	}

	appName := args.Domain.ApplyT(func(s string) string {
		return strings.Join([]string{"app", s}, ".")
	}).(pulumi.StringOutput)

	resource.ConsoleRecord, err = route53.NewRecord(ctx, fmt.Sprintf("%s-console-record", name), &route53.RecordArgs{
		ZoneId: zone.Id(),
		Name:   appName,
		Type:   pulumi.String("A"),
		Aliases: &route53.RecordAliasArray{
			route53.RecordAliasArgs{
				Name:                 args.PublicLoadBalancerDnsName,
				ZoneId:               args.PublicLoadBalancerZoneId,
				EvaluateTargetHealth: pulumi.Bool(true),
			},
		},
	}, options...)

	if err != nil {
		return nil, err
	}

	if args.EnablePrivateLoadBalancerAndLimitEgress {
		apiInternalName := args.Domain.ApplyT(func(s string) string {
			return strings.Join([]string{"api-internal", s}, ".")
		}).(pulumi.StringOutput)

		resource.ApiInternalRecord, err = route53.NewRecord(ctx, fmt.Sprintf("%s-api-internal-record", name), &route53.RecordArgs{
			ZoneId: zone.Id(),
			Name:   apiInternalName,
			Type:   pulumi.String("A"),
			Aliases: &route53.RecordAliasArray{
				route53.RecordAliasArgs{
					Name:                 args.InternalLoadBalancerDnsName,
					ZoneId:               args.InternalLoadBalancerZoneId,
					EvaluateTargetHealth: pulumi.Bool(true),
				},
			},
		}, options...)

		if err != nil {
			return nil, err
		}
	}

	return &resource, nil
}

type ApplicationDnsArgs struct {
	Region                                  string
	EnablePrivateLoadBalancerAndLimitEgress bool
	Domain                                  pulumi.StringOutput
	ZoneName                                pulumi.StringOutput
	PublicLoadBalancerDnsName               pulumi.StringOutput
	PublicLoadBalancerZoneId                pulumi.StringOutput
	InternalLoadBalancerDnsName             pulumi.StringOutput
	InternalLoadBalancerZoneId              pulumi.StringOutput
}

type ApplicationDns struct {
	pulumi.ResourceState

	ApiRecord         *route53.Record
	ApiInternalRecord *route53.Record
	ConsoleRecord     *route53.Record
}

File: ecs-hosted/ts/dns/applicationDns.ts
import * as pulumi from "@pulumi/pulumi";
import * as route53 from "@pulumi/aws/route53";

import { DnsRecordArgs } from "./types";

const namespace = "pulumi:DnsRecord";

export class ApplicationDns extends pulumi.ComponentResource {

    public readonly apiFqdn: pulumi.Output<string>;
    public readonly consoleFqdn: pulumi.Output<string>;

    constructor(name: string, args: DnsRecordArgs, opts?: pulumi.ComponentResourceOptions) {
        super(namespace, name, args, opts);

        const options = pulumi.mergeOptions(opts, { parent: this });

        // retrieve the hosted zone that will host our A records
        // zone should match the domain name.
        const zoneId = pulumi.output(route53.getZoneOutput({
            name: args.zoneName
        }, options)).apply(result => result.zoneId);

        const apiRecord = new route53.Record(`${name}-api-record`, {
            zoneId: zoneId,
            name: pulumi.interpolate `api.${args.domain}`,
            type: "A",
            aliases: [{
                name: args.apiLoadBalancerDnsName,
                zoneId: args.apiLoadBalancerZoneId,
                evaluateTargetHealth: true
            }]
        }, options);

        const consoleRecord = new route53.Record(`${name}-console-record`, {
            zoneId: zoneId,
            name: pulumi.interpolate `app.${args.domain}`,
            type: "A",
            aliases: [{
                name: args.consoleLoadBalancerDnsName,
                zoneId: args.consoleLoadBalancerZoneId,
                evaluateTargetHealth: true
            }]
        }, options);

        this.apiFqdn = apiRecord.fqdn;
        this.consoleFqdn = consoleRecord.fqdn;
    }
}